Pet Patch 1.0: Amazing fun times with pets

Phatcat opened the pull request:
## Hunter's pets
- Checked the attack speed for all tamable beasts according to [this](http://web.archive.org/web/20061112230516/http://tkasomething.com/attackspeeds.php) and corrected them. I of course did not touch those that already had the correct value according to the list.
- Removed data from pet_levelstats that will be irrellevent with pet patch 1.0 once that gets merged.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/classicdb/database/829)

<!-- Reviewable:end -->

cala commented:
Thank you for this pull request.
I had a few comments after looking quickly at it, but I see you fixed most of them recently (the removal of pets stats for Silithid Swarm as an example). The only one that still comes to mind is setting to 0 the melee damage values for a bunch of NPCs: I'm not sure this is right and that core will use 0 values. Chances are good that they will be changed to 1. If you are aiming to create non-melee NPCs, it would be wiser to use ACID for that, I think.

I will review very soon that PR more in details. Thanks again. :+1:

Tobschinski commented on updates/13xx_petguardians_update.sql:
```diff
@@ -0,0 +1,33 @@
+-- I am fixing the core to handle guardians by their values, check https://github.com/cmangos/mangos-classic/pull/116 for reference.
+
+-- Fixed Health of Timbermaw Ancestor (15720)
+UPDATE creature_template SET MinLevelHealth=500, MaxLevelHealth=500 WHERE Entry=15720;
```
A health multiplier of 0.234082397 is rather unlikely, where did you get that health value from?

Phatcat replied:
http://wowwiki.wikia.com/wiki/Defender_of_the_Timbermaw?oldid=289644
Didn't pull it from my rear end, that's for sure ;)

cala replied:
We trust you that you did not invent those values but took them from various sources. :smile:
This brings the usual topic of sources reliability. One can find many sources for WoW content on the Internet or in books. Some are more reliable than other and it also depends on when said source was edited.

Usually, my trust priority is:
- Sniffs (correct patch > later patch)
- Official WoW Bestiary / Guides books
- Screenshots/Videos that have a valid timestamp
- WoW wiki / WoW pedia when cross checked with other sources
- wowhead and the like

And even with sniffs (that are Blizz actual values), we often misunderstand them. So _careful_ is a keyword here.

Regarding Timbermaw Ancestor, I fixed the NPC some time ago when I first noticed an issue with it while testing the pet patch from Cyberium. I used Bestiary values for it and the NPC stats are OK now. The 500 HP value in wow wiki are more than probably an estimation made by the player due to the fact that in 2006, players did not have access to the real HP values (they were not sent to the client). Thus, HP and other values were estimated by players or addons. The correct value for Timbermaw Ancestor is 427 HP. Usually, when seeing a nice flat value for HP, mana or damage, chances are good this is a estimation, not the actual value.

Phatcat commented:
@cala if I want a guardianpet to use EventAI, it should be set in the creature_template table, right?
Also, can I set any creature I want to stand still and just cast spells as TotemAI? Cause then maybe target dummies should be set as such - sorry to be asking in a comment, dunno how else to reach you really.

cala commented:
Thank you for the recent updates. :+1: 

I agree with the attack speed updates. I have to look, but I may have more to add. :smile: 

However, I disagree with the removal of any pet_level_stats. The stats exists for those NPCs because there is at least one spell that summons those NPCs as pet and not as guardian. A good example would be the various Silithid NPCs with the little swarmers. They use spell 10722 at spawn which summons a Silithid Swarm (4196) with `SPELL_EFFECT_SUMMON_PET`. When engaged in combat, they periodically cast spell 6589 which also summons a Silithid Swarm but with `SPELL_EFFECT_SUMMON_GUARDIAN`.

As for the damage stats set to zero, I think this is probably a hack and the proper way to do it would be to use EventAI to prevent the NPCs to move/engage in melee combat.
This reminds me of your question on IRC: if a creature_template has an empty field for AIName, this means that there is no entry for it in creature_ai_scripts table and thus they need to be written/added/backported. A list of the NPCs currently missing an AI is found in this issue: #740 
We can discuss this later on IRC if you want. :smile:

Phatcat commented:
The removal of pet_levelstats is very much done because the function spell_effect_summon_pet doesn't use pet_levelstats, it uses initstatsforlevel - initstatsforlevel uses pet_levelstats, and I am rewriting that one in [this pr](https://github.com/cmangos/mangos-classic/pull/124) - which is why I said those changes shouldn't be applied till (if) that pr gets merged. In the same pr I also rewrote the summon_pet function to make summons default to guardians; [It's done here](https://github.com/cmangos/mangos-classic/pull/124/files#diff-6e2ad27531ee8eb7bafce8a41909816aR3018)

Attack speed updates are still WIP as doing the 2.0s seem like a nightmare given my current method of ctrl+f , so I skipped those initially. Have returned and done 1/10 locally - will push when completely done with them, then you can double check =)

I agree that setting damage to 0 for the npcs would not solves the issues with them, however the reason for that change is that they should not even be able to melee, so why would we keep some random values someone once gave them? - is there any reason they should not be defaulted to 0? as that is the default for entries not having values in those fields, if I am not mistaken?

I have not taken a look at any AI yet because the PetAI needs a bit of work (defensive is not good), and I am not sure petAI nor eventAI is suited for guardian pets? Given that nearly all guardian pets are supposed to do their own thing (eventAI), while also following and defending their owner (petAI).

Phatcat commented:
@cala I would love to get in touch with you, hit me up on IRC any time.

Phatcat commented:
@cala I have corrected this to not include any flags or 0 values for guardians, but only include the things relevant to pet patch 1.0 for the core (and just a few guardian corrections). That way these fit nicely together as a collected update.

cala commented:
Thanks. I assume that it is best to use also that PR to test your patch core-side?

Phatcat commented:
Well, if you want to be sure that I pulled the pet_levelstats out without breaking anything, then you should, but the core patch is not in any way dependant on the db patch, only the removal of pet_levelstats in the db patch is dependant on the core patch, that's all.

Phatcat commented:
@cala here's the table =)

edit: removed everything relating to guardians (still got it locally) - will do a pr on some proper guardian updates once I get the ball rolling on Pet Patch 2.0 - that will however at least be after Engineering 2.0: The Tesla Tragedy (Gnomes Ahoy)

Phatcat commented:
@cala - someone just came in and schooled me:

> What the hell is this... Pet modifiers are spells, called Tamed Pet Passive (DND).
> For example 17206 gives 7% bonus damage to bats, 17208 gives -9% damage, 5% armor and 8% HP to bears. This kind of handling is wrong.

https://github.com/cmangos/issues/issues/931

Table and stat modifiers now removed.

cala commented:
Yes, I saw the notification on core. I did not know or I would have lead you that way, of course. In case he does not provide you the spell IDs, I can try to look for them.

Phatcat commented:
I hope I won't cause too many issues for people now that I dropped those 2 commits... I have added the comment to pet issues, so I can work it into pet patch 2.0

cala commented:
Thanks, I merged the attack speed part now the core supports this. :+1: 
I see you removed the small update file that was fixing stats for one NPC (I don't remember which one) and the experience multiplier for another one. Is this intended?

Regarding the `pet_levelstats` table, I'm still reluctant to drop those entries:
- they are not loaded in core at startup so won't decrease performance if kept
- the whole table is quite small (55 kB) so it's not taking too much space
- the values seems to be good so they may be valuable later to test/define/check an improved stats system (it is hard enough to get reliable data for the DB, I prefer avoid dropping them needlessly)

Phatcat commented:
@cala I'm sorry, but what do you mean that the values are good? the values are arrived at using a formula. Just give me that formula and I'll use that for guardians... Besides that, why would you want to hold on to data that should not, and probably will not, ever be used for anything ever again?
I'm not saying throw them out completely, but remove them from the current db.

You say dropping them needless, I say keeping them needlessly.

Phatcat commented:
Oh, forgot the other part of your message.

> I see you removed the small update file that was fixing stats for one NPC (I don't remember which one) and the experience multiplier for another one. Is this intended?

Yes. Because they are intended to make a dreaded comeback along with all the other guardian changes when I get all those AI kinks worked out..

Phatcat commented:
I'm gonna remove the beast attack speeds from this pr since you've already added those. Then we are down to the question, should this pr get closed or not then? - I doubt we'll ever come to terms any time soon about the removal of that data, so this pr could just end up going stale if left open.

cala commented:
> @cala I'm sorry, but what do you mean that the values are good? the values are arrived at using a formula. Just give me that formula and I'll use that for guardians... Besides that, why would you want to hold on to data that should not, and probably will not, ever be used for anything ever again?

I said that they _seem_ good. We can discuss the reason of the emphasis on that verb on IRC. We do not have a formula or of course we would not need them, but in the eventuality we get a formula one day, I think these data could be a good set for testing. You are right we do not need to keep them in the DB for that specific purpose. I don't have the time to take a decision about that right now. I'll come back to you on this later. :)

Phatcat commented:
> We do not have a formula

Is why I made one; let's see how well it fits =)

~~Okay, leaving this pr open just for good meassure, then.~~
Gonna do the exact opposite instead.

cala commented:
> ~~Okay, leaving this pr open just for good meassure, then.~~
> Gonna do the exact opposite instead.

I lol'd. :laughing:
